src/pages/document.ejs

```diff
@@ -9,7 +9,7 @@
   <link rel="icon" href="<%= context.config.publicPath +'favicon.ico'%>" type="image/x-icon">
   
   <!-- runtime endpoints config -->
-  <script src="<%= context.config.publicPath +'endpoints.js'%>"></script>
+  <script src="<%= context.config.publicPath + context.config.define.ENDPOINTS_ENV%>"></script>
```

Comment:
FYI, I'm working on turning this into something like:

```
  <script type="text/javascript">
    fetch(`${window.origin}/api/v1/endpoints`)
      .then(function(response) {
        if (!response.ok) {
          console.log(`ERROR: ${response.statusText}: ${response.statusText}`);
        } else {
          response.json().then(function(data) {
            window.endpoints = data;
          })
        }
      });
  </script>
```

Although I'm not convinced that's the right packaging and I need to dump the `console.log` and invert that `if`. (Though this morning I've managed to break it every time I simplify anything.)

Reply:
Sweet! Looking forward to being able to query the endpoint configuration on demand.

Reply:
Actually, I was wondering whether you'd object philosophically to including that sort of code in the "root loader" html file. This is encouraging.

However, since you didn't object outright... this works beautifully on my VM, but I'm having problems in the e2e environment. The `fetch` is returning the full index.html, and the `.json()` is throwing because of the initial `<` character. (That's why I had the `console.log()` statements in here; although to actually nail this down I had to go to `response.text()` and print that before trying to manually jsonify it.) This may be partly because this is running too early, before the request mocking is set up, but getting the page html from the yarn server instead of a 404 still confuses me. Why would it ignore the path? Do you have any ideas here? Is there some yarn configuration, for example, that's making the development server return the page html for _every_ `GET` until we set up the mock overrides???

Reply:
My only concern is the added latency and overhead introduced in querying for a file before loading the dashboard bundle. Since the endpoint config is very small in size we can assume minimal latency and implement strategies for caching the definition in the future. 

Have you tried setting `MOCK=true` in the e2e environment? We currently do this in `package.json` for starting the dashboard in mocked environment: `"start:mock": "cross-env UMI_UI=none MOCK=true umi dev"`. It also may be helpful to debug the environment by setting the e2e tests to run without the headless mode: 

```
browser = await puppeteer.launch({
    headless: true -> false, // headless should be set to false 
    args: ['--no-sandbox'],
  });
```